### ISSUE TITLE
higher-level(feat): Handle {De,En}coding of Mailbox

### ISSUE THREAD

**Issue by duesee** [duesee/imap-codec#370]
Rust MUAs need to handle {de,en}coding of mailbox (i.e., folder) names on their own. This means, that it might be required to use "decode_utf7", "decode_base64", ... before being able to display a mailbox. Ideally, the higher-level library already does that.

Questions:

* Is there a canonical form?
* Can we use `Cow<String>`?
* What happens if we create a mailbox with colliding encodings?
  * -> Not sure, but at least the RFC says "Modified BASE64 MUST NOT be used to represent any printing US-ASCII character which can represent itself." Sounds canonical to me?

**Comment by duesee** [comment]
From my reading of RFC 3501 (IMAP4rev1), modified UTF-7 is the only thing we need to care about in mailbox names.

RFC 9051 (IMAP4rev2) says:

> Support for the Mailbox International Naming Convention described in this section is not required for IMAP4rev2-only clients and servers. It is only used for backward compatibility with IMAP4rev1 implementations.

Encoding is also relevant in an `Envelope` (e.g., in an email subject). Here we can have, e.g., `=?UTF-8?Q?<data>?=`, `=?base64?Q?<data>?=`, ...

**Comment by soywod** [comment]
It is exactly what I used to do with `imap` (and what I am doing right now with `imap-flow`). It could be great to integrate encoding/decoding inside the lib!

**Comment by duesee** [comment]
> Ultimately this can lead to clients sending invalid data to the server, which is not very desirable.

Totally agree. If `imap-types` allows this, I consider it a bug.

It would help if you could provide a concrete mailbox value that leads to the server error you mentioned in the imap-client repository.

According to IMAP's formal syntax, even a mailbox value of, e.g., "h²llo" is allowed when encoded as a literal, i.e., ...

```imap
A SELECT {6}/r/n
H²llo
```

For reference, the ABNF for mailbox is ...

```abnf
mailbox         = "INBOX" / astring

astring         = 1*ASTRING-CHAR / string

string          = quoted / literal

literal         = "{" number "}" CRLF *CHAR8
                    ; Number represents the number of CHAR8s

CHAR8           = %x01-ff
                    ; any OCTET except NUL, %x00
```

The formal syntax is the reason why `imap-types` allows to construct a mailbox from almost any byte-sequence (except those containing a NULL due to how CHAR8 is defined).

On the other hand, the RFC 3501 tells in prose...

> Mailbox names are 7-bit.

I start to remember why I try to avoid this topic :P

**Comment by duesee** [comment]
Ah, just saw that I explained the "automatic construction" here: https://github.com/duesee/imap-codec/tree/main/imap-types#examples Maybe it's useful to someone reading this ticket.

**Comment by duesee** [comment]
Yeah. I think you are on the right track here. Let's tackle this finally.

**Comment by jakoschiko** [comment]
Would it make sense to provide the supported extensions to the constructor of `Mailbox`? Then the `Mailbox` would be valid or invalid depending on the context (the supported extensions).

The `Codec`s of `imap-codec` could also contain the supported extensions and use them for decoding.

**Comment by duesee** [comment]
Yes, exactly. See, e.g., https://github.com/duesee/imap-codec/issues/239.
